Improve mocking code snippets for better readability (#1049)
ballerina-platform/ballerina-dev-website: swan-lake/learn/mocking.md

Let's make changes to the example in the [Quick Start](/swan-lake/learn/testing-quick-start) page. In order to test the
 `getRandomJoke` function without actually calling the endpoint, use the `test:mock` function to mock the `get` remote function of the client object.

Change the content of the test file as follows:

> This is the definition of the mock object which can be used as the test double for the `clientEndpont` object. Only the `get` function is implemented since it is the only function used in the sample.

Reviewer: ```suggestion
This is the definition of the mock object, which can be used as the test double for the `clientEndpont` object. Only the `get` function is implemented since it is the only function used in the sample.
```

Attempting to call any other function will result in a runtime exception.

```ballerina
import ballerina/http;
 
public type MockHttpClient client object {
    public remote function get(@untainted string path, public http:RequestMessage message = ()) 
    	returns http:Response|http:ClientError {

        http:Response response = new;
        response.statusCode = 500;
        return response;
    }
};
```

> Now an instance of this object can be used for testing the `getRandomJoke` function.

Reviewer: ```suggestion
Now, an instance of this object can be used for testing the `getRandomJoke` function.
```

Instead of creating a test double, you may also choose to create a default mock object and stub the functions to return a specific value or to do nothing.

The example in the [Quick Start](/swan-lake/learn/testing-quick-start) page shows how the `get` function of the client object can be

>  stubbed to return a value. Let’s make changes to that example to get a random joke from a specific category (e.g. food, movies).

Reviewer: ```suggestion
 stubbed to return a value. Let’s make changes to that example to get a random joke from a specific category (e.g., food or movies).
```

**main.bal**

```ballerina
import ballerina/io;
import ballerina/http;
import ballerina/stringutils;

http:Client clientEndpoint = new("https://api.chucknorris.io/jokes/");

// This function performs a `get` request to the Chuck Norris API and returns a random joke 
// or an error if API invocations fail.
function getRandomJoke(string name, string category = "food") returns string|error {
    http:Response response = checkpanic clientEndpoint->get("/categories");

    // check if the provided category is available
    if (response.statusCode == http:STATUS_OK) {
        json[] categories = <json[]>response.getJsonPayload();
        if (!isCategoryAvailable(categories, category)) {
            error err = error("'" + category + "' is not a valid category.");
            io:println(err.message());
            return err;
        }
    } else {
    	return createError(response);
    }

>     // get random joke from the provided category

Reviewer: ```suggestion
    // get a random joke from the provided category
```

}

// Returns an error based on the HTTP response
function createError(http:Response response) returns error {
    error err = error("error occurred while sending GET request");
    io:println(err.message(), ", status code: ", response.statusCode);
    return err;
}
```
#### Stubbing to return a specific value

***Example:***

The below util functions are used to construct mock responses required for testing.
  
  ```ballerina
import ballerina/http;

> // Returns a mock HTTP response to be used for random joke API invocation

Reviewer: ```suggestion
// Returns a mock HTTP response to be used for a random joke API invocation
```

function getMockResponse() returns http:Response {
    http:Response mockResponse = new;
    json mockPayload = {"value":"When Chuck Norris wants an egg, he cracks open a chicken."};
    mockResponse.setPayload(mockPayload);
    return mockResponse;
}

// Returns a mock response to be used for the category API invocation
function getCategoriesResponse() returns http:Response {
    http:Response categoriesRes = new;
    json[] payload = ["animal","food","history","money","movie"];
    categoriesRes.setJsonPayload(payload);
    return categoriesRes;
}
```

**main_test.bal**
 

>  1. This test stubs the behavior of the `get` function to return a specific value function

Reviewer: ```suggestion
 1. This test stubs the behavior of the `get` function to return a specific value function.
```

 
```ballerina
import ballerina/test;
import ballerina/http;

@test:Config {}
public function testGetRandomJoke() {
    // create a default mock HTTP Client and assign it to the `clientEndpoint` object
    clientEndpoint = <http:Client>test:mock(http:Client);

>     // stub to return the specified mock response when `get` function is called

Reviewer: ```suggestion
    // stub to return the specified mock response when the `get` function is called
```

    test:prepare(clientEndpoint).when("get").thenReturn(getMockResponse());

    // stub to return the specified mock response when the specified argument is passed
    test:prepare(clientEndpoint).when("get").withArguments("/categories").thenReturn(getCategoriesResponse());

    // invoke the function to test
    string result = checkpanic getRandomJoke("Sheldon");

    // verify the return value against the expected string
    test:assertEquals(result, "When Sheldon wants an egg, he cracks open a chicken.");
}
```

> 2. This test stubs the behavior of the `get` function to return a specified sequence of values for each `get` function invocation. 

Reviewer: ```suggestion
2. This test stubs the behavior of the `get` function to return a specified sequence of values for each `get` function invocation
```

i.e., the first call to `get` will return the fist argument and the second call will return the second argument.

```ballerina
import ballerina/test;
import ballerina/http;

@test:Config {}
public function testGetRandomJoke() {
    // create a default mock HTTP Client and assign it to the `clientEndpoint` object
    clientEndpoint = <http:Client>test:mock(http:Client);

    // stub to return the corresponding value for each invocation 
    test:prepare(clientEndpoint).when("get").thenReturnSequence(getCategoriesResponse(), getMockResponse());

    // invoke the function to test
    string result = checkpanic getRandomJoke("Sheldon");

    // verify the return value against the expected string
    test:assertEquals(result, "When Sheldon wants an egg, he cracks open a chicken.");
}
```

> 3. This test shows how to stub a member variable value of the `clientEndpoint` object

Reviewer: ```suggestion
3. This test shows how to stub a member variable value of the `clientEndpoint` object.
```

```

**main_test.bal**
 
 This is the initialization of the mock function that should be called in place of the `intAdd` function

```ballerina
import ballerina/test;

@test:MockFn { functionName: "intAdd" }
test:MockFunction intAddMockFn = new();
```

> After the initialization, the following different options can be used to stub the behavior of a function written in the

Reviewer: ```suggestion
After the initialization, the following options can be used to stub the behaviour of a function written in the
```

>  module that is under test.

Reviewer: ```suggestion
 module that is under testing.
```

Author: Fixed in  b7e04c6
